lib/action_controller/serialization.rb

```diff
@@ -18,46 +19,49 @@ def serialization_scope
         respond_to?(_serialization_scope, true)
     end
 
+    # Deprecated
     def get_serializer(resource)
-      @_serializer ||= @_serializer_opts.delete(:serializer)
-      @_serializer ||= ActiveModel::Serializer.serializer_for(resource)
-
-      if @_serializer_opts.key?(:each_serializer)
-        @_serializer_opts[:serializer] = @_serializer_opts.delete(:each_serializer)
-      end
-
-      @_serializer
+      fail ActiveModel::Serializer::Error, "ActionController::Serialization##{__method__} has been removed. "\
+        "Please use ActiveSupport::Serializer#build instead"
     end
 
+    # Deprecated
     def use_adapter?
-      !(@_adapter_opts.key?(:adapter) && !@_adapter_opts[:adapter])
+      fail ActiveModel::Serializer::Error, "ActionController::Serialization##{__method__} has been removed. "\
+        "Please use ActiveSupport::Serializer#build instead"
     end
 
     [:_render_option_json, :_render_with_renderer_json].each do |renderer_method|
       define_method renderer_method do |resource, options|
-        @_adapter_opts, @_serializer_opts =
-          options.partition { |k, _| ADAPTER_OPTION_KEYS.include? k }.map { |h| Hash[h] }
-
-        if use_adapter? && (serializer = get_serializer(resource))
-
-          @_serializer_opts[:scope] ||= serialization_scope
-          @_serializer_opts[:scope_name] = _serialization_scope
-
-          # omg hax
-          object = serializer.new(resource, @_serializer_opts)
-          adapter = ActiveModel::Serializer::Adapter.create(object, @_adapter_opts)
-          super(adapter, options)
-        else
-          super(resource, options)
+        serializable_resource = ActiveModel::Serializer::build(resource, options) do |builder|
+          if builder.serializer?
+            builder.serialization_scope ||= serialization_scope
+            builder.serialization_scope_name = _serialization_scope
+            builder.adapter
+          else
+            resource
```

Comment:
However, I think that we should try to use the builder interface as much as possible in specs from now on, passing in :adapter, :serializer, :each_serializer, etc., and only instantiating adapters and serializers directly in builder tests (which don't yet exist of course), and possibly in serializer tests.  I think it would be a big win if we could narrow the necessary/recommended/whatever public interface to the builder and maybe something else, but I think passing `adapter: false` would still be better in those cases.